### PR TITLE
add parse and serialize methods in the call_request message.

### PR DESCRIPTION
**Pull request by junchaowu**
for dev purpose

### CODE REVIEW

File: python/tchannel/messages/call_request.py [hunk 1]
```diff
@@ -36,6 +36,7 @@ class CallRequestMessage(BaseMessage):
 
     FLAGS_SIZE = 1
     TTL_SIZE = 4
+    # span_id + parent_id + trace_id = 3*8 = 24
```
Comment (abhinav) [inline python/tchannel/messages/call_request.py]: ?

File: python/tchannel/messages/call_response.py [hunk 2]
```diff
@@ -27,3 +32,87 @@ class CallResponseMessage(CallRequestMessage):
         'arg_2',
         'arg_3',
     )
+
+    FLAGS_SIZE = 1
+    CODE_SIZE = 1
+    # span_id + parent_id + trace_id = 3*8 = 24
```
Comment (abhinav) [inline python/tchannel/messages/call_response.py]: I don't understand what this is for

File: python/tchannel/messages/call_response.py [hunk 3]
```diff
@@ -27,3 +32,87 @@ class CallResponseMessage(CallRequestMessage):
...
+
+        self.headers, _ = self._read_headers(
+            payload,
+            self.NH_SIZE,
+            self.HEADER_SIZE,
+        )
+
+        self.checksum_type = read_number(payload, self.CSUMTYPE_SIZE)
+        if self.checksum_type:
+            self.checksum = read_number(payload, self.CSUM_SIZE)
```
Comment (abhinav) [inline python/tchannel/messages/call_response.py]: Checksum size is not exactly to 4 bytes. It just so happens that all supported checksum types are 32-bit. use a mapping from checksum_type to checksum size (might as well update call req while you're at it).

File: python/tchannel/messages/call_response.py [hunk 4]
```diff
@@ -27,3 +32,87 @@ class CallResponseMessage(CallRequestMessage):
...
+
+        # TODO check if we're at the end of the stream by counting bytes
```
Comment (abhinav) [inline python/tchannel/messages/call_response.py]: What?

File: python/tchannel/messages/call_response.py [hunk 5]
```diff
@@ -27,3 +32,87 @@ class CallResponseMessage(CallRequestMessage):
...
+        self.arg_1, _ = read_variable_length_key(
+            payload,
+            self.ARG_SIZE,
```
Comment (abhinav) [inline python/tchannel/messages/call_response.py]: This is not really arg size. It's arg length size.

File: python/tchannel/messages/call_response.py (hunk 3 above)
Comment (junchaowu) [inline python/tchannel/messages/call_response.py]: done

File: python/tchannel/messages/call_request.py [hunk 6]
```diff
@@ -34,17 +37,24 @@ class CallRequestMessage(BaseMessage):
...
     TTL_SIZE = 4
+    # sizeOf(span_id) + sizeOf(parent_id) + sizeOf(trace_id) = 3*8 = 24
     TRACE_SIZE = 8
     TRACEFLAGS_SIZE = 1
     SERVICE_SIZE = 2
     NH_SIZE = 1
     HEADER_SIZE = 1
     CSUMTYPE_SIZE = 1
-    CSUM_SIZE = 4
+    CSUM_SIZE = CHECKSUM[CSUMTYPE_SIZE]
```
Comment (sectioneight) [inline python/tchannel/messages/call_request.py]: I don't think you can do this at import time -- it's going to vary based on the message.

File: python/tchannel/messages/call_response.py [hunk 7]
```diff
@@ -27,3 +35,95 @@ class CallResponseMessage(CallRequestMessage):
...
+    def parse(self, payload, size):
+        """Parse a call request message from a payload."""
+        self.flags = read_number(payload, self.FLAGS_SIZE)
+        self.code = read_number(payload, self.CODE_SIZE)
+
+        self.span_id = read_number(payload, self.TRACE_SIZE)
+        self.parent_id = read_number(payload, self.TRACE_SIZE)
+        self.trace_id = read_number(payload, self.TRACE_SIZE)
+
+        self.traceflags = read_number(payload, self.TRACEFLAGS_SIZE)
```
Comment (sectioneight) [inline python/tchannel/messages/call_response.py]: This looks copy-pasted from above. We should avoid copying code at all costs. It's generally an indication that some refactoring / generalization is in order.

File: python/tchannel/messages/call_response.py [hunk 8]
```diff
@@ -27,3 +35,95 @@ class CallResponseMessage(CallRequestMessage):
...
+
+    def serialize(self, out):
+        """Write a call request message out to a buffer."""
+        out.extend(write_number(self.flags, self.FLAGS_SIZE))
+        out.extend(write_number(self.code, self.code_SIZE))
+
+        out.extend(write_number(self.span_id, self.TRACE_SIZE))
+        out.extend(write_number(self.parent_id, self.TRACE_SIZE))
+        out.extend(write_number(self.trace_id, self.TRACE_SIZE))
+        out.extend(write_number(self.traceflags, self.TRACEFLAGS_SIZE))
```
Comment (sectioneight) [inline python/tchannel/messages/call_response.py]: Some comment: please don't copy paste code around

File: python/tchannel/messages/call_request.py (hunk 6 above)
Comment (junchaowu) [inline python/tchannel/messages/call_request.py]: oh you are right.

File: python/tchannel/messages/call_response.py (hunk 7 above)
Comment (junchaowu) [inline python/tchannel/messages/call_response.py]: this is just the method in the subclass overrides the base class method. There are some difference between subclass one and base class one.

File: python/tchannel/messages/call_response.py (hunk 8 above)
Comment (junchaowu) [inline python/tchannel/messages/call_response.py]: same comments.

File: python/tchannel/messages/call_response.py (hunk 7 above)
Comment (junchaowu) [inline python/tchannel/messages/call_response.py]: These two method did almost same thing but for different payload format.

Comment (sectioneight) [inline python/tchannel/messages/call_response.py]: I get that there are some parts that are different, which is why I comment on this part -- reading the zipkin and traceflags info. I feel pretty strongly that outside of tests, we should not have the exact same code copy/pasted into multiple files.

Comment (sectioneight) [inline python/tchannel/messages/call_response.py]: In other words -- extract the span_id, parent_id, trace_id, traceflags stuff (and whatever else is in common) into a method.

File: python/tchannel/messages/call_response.py (hunk 8 above)
Comment (sectioneight) [inline python/tchannel/messages/call_response.py]: Same response: extract code that is performing the exact same calls into a re-usable method.

File: python/tchannel/messages/call_response.py [hunk 9]
```diff
@@ -27,3 +35,95 @@ class CallResponseMessage(CallRequestMessage):
...
+            self.arg_2,
+            self.ARG_LENGTH,
+            encode=False,
+        )
+        write_variable_length_key(
+            out,
+            self.arg_3,
+            self.ARG_LENGTH,
+            encode=False,
+        )
```
Comment (sectioneight) [inline python/tchannel/messages/call_response.py]: This also looks extractable

**Comment by blampe** [comment]
lgtm

**Pull request merged**